v-next/hardhat/src/internal/builtin-plugins/solidity/build-results.ts

```diff
@@ -54,27 +56,95 @@ export function throwIfSolidityBuildFailed(
   }
 }
 
+/**
+ * This function returns the build infos and outputs associated with the given
+ * Solidity build results.
+ *
+ * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
+ * @returns The build infos in the Hardhat v3 format as expected by the EDR.
+ */
+export async function getBuildInfos(
+  results: SuccessfulSolidityBuildResults,
+  artifactManager: ArtifactManager,
+): Promise<BuildInfoAndOutput[]> {
+  const buildIds = Array.from(new Set(results.values())).map(
+    ({ buildId }) => buildId,
+  );
+
+  return Promise.all(
+    buildIds.map(async (buildId) => {
+      const buildInfoPath = await artifactManager.getBuildInfoPath(buildId);
+      const buildInfoOutputPath =
+        await artifactManager.getBuildInfoOutputPath(buildId);
+
+      assertHardhatInvariant(
+        buildInfoPath !== undefined,
+        "buildInfoPath should not be undefined",
+      );
+      assertHardhatInvariant(
+        buildInfoOutputPath !== undefined,
+        "buildInfoOutputPath should not be undefined",
+      );
+
+      const buildInfo = await readBinaryFile(buildInfoPath);
+      const output = await readBinaryFile(buildInfoOutputPath);
+
+      return {
+        buildInfo,
+        output,
+      };
+    }),
+  );
+}
+
 /**
  * This function returns the artifacts generated during the compilation associated
  * with the given Solidity build results. It relies on the fact that each successful
  * build result has a corresponding artifact generated property.
+ *
+ * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
+ * @returns The artifacts in the format expected by the EDR.
  */
 export async function getArtifacts(
   results: SuccessfulSolidityBuildResults,
-  artifactsRootPath: string,
+  buildInfos: BuildInfoAndOutput[],
 ): Promise<EdrArtifact[]> {
-  const artifacts: EdrArtifact[] = [];
+  const solcVersions = Object.fromEntries(
+    buildInfos.map(({ buildInfo }) => {
+      const solidityBuildInfo: SolidityBuildInfo = JSON.parse(
+        new TextDecoder("utf-8").decode(buildInfo),
+      );
+
+      return [solidityBuildInfo.id, solidityBuildInfo.solcVersion];
+    }),
+  );
```

Comment:
We need the build ID to solc version mapping in order to create artifacts in the format EDR expects.

I decided to retrieve this mapping from the build infos, which we have to create and pass over to EDR anyway. By doing so, I was able to avoid reading the build info file from disk for the second time.